openmc/cmfd.py

```diff
@@ -0,0 +1,3072 @@
+            # and energy bin and updates its weight
+            openmc.lib.source_bank()['wgt'] *= self._weightfactors[
+                    mesh_ijk[:,0], mesh_ijk[:,1], mesh_ijk[:,2], energy_bins]
+
+            if openmc.lib.master() and np.any(source_energies < energy[0]):
+                print(' WARNING: Source pt below energy grid')
+                sys.stdout.flush()
+            if openmc.lib.master() and np.any(source_energies > energy[-1]):
+                print(' WARNING: Source pt above energy grid')
+                sys.stdout.flush()
+
+    def _count_bank_sites(self):
+        """Determines the number of fission bank sites in each cell of a given
+        mesh and energy group structure.
+        Returns
+        -------
+        bool
+            Wheter any source sites outside of CMFD mesh were found
+
+        """
+        # Initialize variables
+        m = openmc.lib.meshes[self._mesh_id]
+        bank = openmc.lib.source_bank()
+        energy = self._egrid
+        sites_outside = np.zeros(1, dtype=bool)
+        nxnynz = np.prod(self._indices[0:3])
+        ng = self._indices[3]
+
+        outside = np.zeros(1, dtype=bool)
+        self._sourcecounts = np.zeros((nxnynz, ng))
+        count = np.zeros(self._sourcecounts.shape)
+
+        # Get location and energy of each particle in source bank
+        source_xyz = openmc.lib.source_bank()['r']
+        source_energies = openmc.lib.source_bank()['E']
+
+        # Convert xyz location to mesh index and ravel index to scalar
+        mesh_locations = np.floor((source_xyz - m.lower_left) / m.width)
+        mesh_bins = mesh_locations[:,2] * m.dimension[1] * m.dimension[0] + \
+            mesh_locations[:,1] * m.dimension[0] + mesh_locations[:,0]
+
+        # Check if any source locations lie outside of defined CMFD mesh
+        if np.any(mesh_bins < 0) or np.any(mesh_bins >= np.prod(m.dimension)):
+            outside[0] = True
+
+        # Determine which energy bin each particle's energy belongs to
+        # Separate into cases bases on where source energies lies on egrid
+        energy_bins = np.zeros(len(source_energies), dtype=int)
+        idx = np.where(source_energies < energy[0])
+        energy_bins[idx] = 0
+        idx = np.where(source_energies > energy[-1])
+        energy_bins[idx] = ng - 1
+        idx = np.where((source_energies >= energy[0]) &
+                       (source_energies <= energy[-1]))
+        energy_bins[idx] = np.digitize(source_energies, energy) - 1
+
+        # Determine all unique combinations of mesh bin and energy bin, and
+        # count number of particles that belong to these combinations
+        idx, counts = np.unique(np.array([mesh_bins, energy_bins]), axis=1,
+                                return_counts=True)
+
+        # Store counts to appropriate mesh-energy combination
+        count[idx[0].astype(int), idx[1].astype(int)] = counts
+
+        if have_mpi:
```

Comment:
Looks like you are aware of this, but definitely take care in babysitting MPI clients when running high-throughput calculations in clusters. In practice, I've found even mid-size MPI cluster deployments to be very irritable and can die easily, even when doing state checkpointing.